경북대 BE 박신영 STEP1: 카카오 로그인

parknew0 opened the pull request:
안녕하세요 멘토님! 1단계 카카오 로그인 구현 완료 후 PR 요청드립니다!

## 구현 내용
- OAuth 2.0 (Authorization Code Grant): 카카오 로그인 공식 문서를 기반으로 인증 흐름을 구현했습니다.

- 보안: API Secret Key 등 민감 정보는 .gitignore 처리된 application-oauth.properties 파일로 분리하여 관리하도록 설정했습니다.

- 사용자 식별: 카카오로부터 사용자의 이메일을 받아오려고 했으나, 이메일 수집을 하려면 추가 기능을 신청해야해서, 이메일 대신 카카오로부터 받은 고유 회원번호(ID)를 기준으로 사용자를 식별하고 가입/로그인 처리합니다.

- 신규 유저의 경우 kakaoId@kakao.user 형태의 가상 이메일을 생성하여 DB에 저장합니다.

- 인증 처리: 로그인 성공 시, 서버는 자체 JWT를 발급합니다. 이 토큰은 중간 페이지(oauth-redirect.html)를 거쳐 브라우저의 localStorage에 저장됩니다.

parknew0 committed:
feat: 초기 코드 마이그레이션

parknew0 committed:
feat(docs): README 1단계 구현기능 목록 업데이트

parknew0 committed:
feat(docs): README 1단계 구현기능 목록 업데이트

parknew0 committed:
feat(auth): 카카오 OAuth 클라이언트 정보 관리 설정

- application-oauth.properties 파일을 생성하여 민감 정보 분리

parknew0 committed:
feat(controller): 카카오 로그인 페이지 리다이렉션 구현

parknew0 committed:
feat(dto): 카카오 API 통신용 DTO 정의

parknew0 committed:
feat(auth): 카카오 인증 콜백 처리 및 액세스 토큰 요청 구현

parknew0 committed:
feat(auth): 액세스 토큰으로 카카오 사용자 정보 조회 기능 구현

parknew0 committed:
feat(auth): 카카오 정보 기반 로그인/회원가입 및 JWT 발급 구현

parknew0 committed:
refactor(auth): 카카오 사용자 식별 기준을 이메일에서 ID로 변경

카카오 이메일 정보는 별도의 검수 절차 없이는 수집할 수 없는 비즈니스 제약사항을 반영.
이에 따라 사용자를 고유하게 식별하는 기준을 이메일에서 카카오 회원번호(ID)로 변경함.
DB email 필드는 `kakaoId + "@kakao.user"` 형태의 가상 이메일을 생성하여 NOT NULL 제약조건을 만족시킴.

parknew0 committed:
fix(auth): 카카오 로그인 후 토큰 JSON이 노출되는 문제 수정

중간 리디렉션 페이지(oauth-redirect.html)를 추가함.
해당 페이지의 스크립트는 토큰을 localStorage에 저장하고 메인 페이지로
리디렉션.

parknew0 committed:
Merge branch 'parknew0' into step1

urarik commented on .gitignore:
```diff
@@ -38,3 +38,6 @@ out/
 .DS_Store
+
+# Secret properties
+application-oauth.properties
```
application-oauth.properties 자체를 숨겨버리면 여기에 있는 프로펄티가 어떤건지 알기가 어려워요. 코드상으로 파악해야 하는데 쉽지 않을 것 같습니다. application.properties에 값을 넣어두는 대신 환경변수로 치환되게 하는 방식은 어떨까요~?

parknew0 replied:
application.properties에서 플레이스홀더를 사용해 환경 변수로부터 설정값을 읽어오도록 수정했습니다. 필요한 환경 변수 목록도 README.md에 함께 정리했습니다!

urarik commented on src/main/java/gift/config/KakaoOauthProperties.java:
```diff
@@ -0,0 +1,37 @@
+import org.springframework.stereotype.Component;
+
+@Component
+@ConfigurationProperties(prefix = "kakao")
+public class KakaoOauthProperties {
```
연관있는 프로퍼티끼리 모아둔 것 좋네요~ 👍

urarik commented on src/main/java/gift/config/KakaoOauthProperties.java:
```diff
@@ -0,0 +1,37 @@
+        return clientId;
+    }
+
+    public void setClientId(String clientId) {
+        this.clientId = clientId;
+    }
```
setter를 설정해야할 필요가 있을까요? 불변객체의 장단점을 한 번 공부해보세요~ 그리고 ConstructorBinding 어노테이션도 추가로 보시면 좋을 것 같습니다!

parknew0 replied:
불변객체는 객체의 상태가 변하지 않으니깐 예측 가능성이 높아지고, 여러 스레드가 동시에 접근해도 안전하네요,
KakaoOauthProperties를 불변 record로 변경하고, @EnableConfigurationProperties를 통해 등록하는 방식으로 수정했습니다!

urarik commented on src/main/java/gift/controller/OAuthController.java:
```diff
@@ -0,0 +1,44 @@
+
+    @GetMapping("/kakao")
+    public void kakaoLogin(HttpServletResponse response) throws IOException {
+        String url = "https://kauth.kakao.com/oauth/authorize?"
+                + "client_id=" + kakaoOauthProperties.getClientId()
+                + "&redirect_uri=" + kakaoOauthProperties.getRedirectUri()
+                + "&response_type=code"
+                + "&scope=talk_message";
```
+를 사용한 방식과 StringBuilder를 사용한 방식의 장단점이 뭘까요~?

parknew0 replied:
`+` 연산은 코드가 간결하고 가독성이 좋은 반면 +연산을 할 때마다 String 객체들이 계속 생성되고 소멸되기 때문에 성능 저하와 메모리 낭비를 유발할 수 있습니다,
`StringBuilder`는 반복문 안에서 문자열을 합칠 때 성능상의 이점이 큰 것으로 알고 있습니다. 
다만 단발성으로 문자열을 합치는 경우, Java 컴파일러가 내부적으로 StringBuilder를 통해 최적화를 해줘서 성능상의 차이점은 없는 것으로 알고 있습니다!

urarik commented on src/main/java/gift/dto/kakao/KakaoTokenResponse.java:
```diff
@@ -0,0 +1,24 @@
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public record KakaoTokenResponse(
+        @JsonProperty("token_type")
```
@JsonNaming에 대해 공부해보세요~

parknew0 replied:
멘토님, 알려주신 @JsonNaming에 대해 학습하고 적용해 보았습니다.
반복되던 @JsonProperty 코드가 사라져서 코드가 훨씬 깔끔하고 명확해졌네요. 좋은 것 알려주셔서 정말 감사합니다!

urarik commented on src/main/java/gift/dto/kakao/KakaoUserInfoResponse.java:
```diff
@@ -0,0 +1,47 @@
+
+    public String getNickname() {
+        if (this.kakaoAccount == null || this.kakaoAccount.profile() == null) {
+            return null;
```
null을 리턴하는 것보다 Optional을 리턴하는 것이 더 좋지 않을까요~?
이 메소드의 사용자는 String만 봐선 null을 생각하지 못할 수 있지만 Optional<String>을 보면 null을 웬만하면 생각할 것이니까요~

parknew0 replied:
알려주신 대로 DTO의 getter 메소드들이 null 대신 Optional을 반환하도록 수정했습니다. 코드가 훨씬 안전하고 세련되게 바뀐 것 같아 마음에 듭니다.

urarik commented on src/main/java/gift/service/MemberService.java:
```diff
@@ -79,4 +79,16 @@ public void saveMember(MemberRequest request) {
+    public Member findOrCreateMemberByKakaoId(Long kakaoId) {
+        return memberRepository.findByKakaoId(kakaoId)
+                .orElseGet(() -> {
+                    String email = kakaoId + "@kakao.user";
+                    String temporaryPassword = java.util.UUID.randomUUID().toString();
```
이 둘을 null로 두는 것과 임의의 값을 넣어두는 것의 장단점이 뭘까요?

parknew0 replied:
임의의 값을 사용하면 DB 스키마 변경 없이 기존 Member 엔티티를 재사용할 수 있는 장점이 있지만, 실제로 사용되지 않는 데이터가 저장된다는 문제가 있고 암묵적인 규칙이 생기는 단점이 있는 것 같습니다.

반면, null을 사용하면 데이터의 의미가 명확해지고 유저 구분이 쉬워지지만, 스키마를 변경해야 하고 코드 전반에 null 처리 로직을 추가해야 하는 부담이 있다는 점을 이해했습니다.

장기적으로는 nullable로 변경하고 null을 사용하는 것이 더 올바른 설계 방향이라고 생각됩니다. null로 두도록 관련 코드를 수정했습니다!

urarik commented on src/main/java/gift/service/OAuthService.java:
```diff
@@ -0,0 +1,91 @@
+
+        HttpEntity<MultiValueMap<String, String>> requestEntity = new HttpEntity<>(body, headers);
+
+        RestTemplate restTemplate = new RestTemplate();
```
RestClient를 사용해보시는건 어떤가요? 
그리고 실제 요청보내는 부분을 KakaoClient 같은 클래스로 분리하는건 어떨까요?

parknew0 replied:
카카오 API 통신을 전담하는 KakaoClient를 새로 생성하여 책임을 분리했습니다. 이제 OAuthService는 KakaoClient에 의존하여 비즈니스 로직에만 집중할 수 있게 되었습니다. OAuthService는 이제 HTTP 통신의 구체적인 방법을 전혀 몰라도 KakaoClient에게 요청만 보내면 되므로 코드가 훨씬 깔끔해지고 각 클래스의 역할이 명확해진 것 같습니다!

urarik replied:
@parknew0 제 생각에는 이 방식이 테스트 작성할 때도 더 좋아질 것 같아요. 왜 그럴까요~?

urarik commented on src/main/java/gift/service/OAuthService.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    private KakaoTokenResponse getKakaoToken(String code) {
+        String url = "https://kauth.kakao.com/oauth/token";
```
restTemplate을 클래스의 멤버변수로 두고 baseUrl을 설정해두는 것은 어떨까요?
그러면 이런 메소드에선 restTemplate.exchange("oauth/token") .... 로 사용할 수 있을 것 같아요~

parknew0 replied:
비효율적인 부분을 짚어주시고 더 나은 설계 방향을 알려주셔서 감사합니다.
덕분에 아이디어를 얻어, RestClientConfig를 만들어 baseUrl이 다른 두 개의 RestClient 빈을 등록하고, KakaoClient에서 이를 주입받아 재사용하도록 수정해볼 수 있었습니다!

urarik reviewed (changes requested):
수고 많으셨습니다~ 코멘트 달아두었으니 확인하고 답변해주세요! 그리고 추가로 테스트 코드도 작성해보세요~

parknew0 committed:
chore(config): 카카오 oauth를 위한 환경 변수 적용

parknew0 committed:
Merge remote-tracking branch 'origin/step1' into step1

# Conflicts:
#	src/main/resources/application.properties

parknew0 committed:
refactor(config): record를 사용해 KakaoOauthProperties를 불변 객체로 변경

parknew0 committed:
refactor(dto): @JsonNaming을 적용하여 Kakao DTO 클래스 간소화

기존에 모든 필드에 반복적으로 사용되던 @JsonProperty 어노테이션을 제거하고, 클래스 레벨에 @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)을 적용

parknew0 committed:
refactor(dto): Optional을 사용해 DTO의 null 안정성 개선

parknew0 committed:
refactor(member): 소셜 로그인 시 email/password에 null을 사용하도록 리팩토링

parknew0 committed:
refactor(oauth): RestClient 사용 및 KakaoClient로 책임 분리

- 카카오 API 통신을 전담하는 KakaoClient를 새로 생성하여 책임을 분리했습니다.
- 이제 OAuthService는 KakaoClient에 의존하여 비즈니스 로직에만 집중합니다.

parknew0 committed:
refactor(client): RestClient 빈 주입 방식으로 외부 API 클라이언트 개선

기존에 KakaoClient 내부에서 RestClient를 매번 생성하던 비효율적인 방식을 개선하여, 애플리케이션 실행 시 생성된 Bean을 재사용하도록 구조를 변경합니다.

- RestClientConfig:
  - `@Configuration`을 사용하여 RestClient 빈 설정 클래스를 생성합니다.
  - `@Bean`과 `@Qualifier`를 사용해 `kauth`, `kapi` 등 용도별로 `baseUrl`이 다른 RestClient 빈을 등록합니다.

- KakaoClient:
  - 생성자를 통해 설정된 RestClient 빈들을 주입받습니다.
  - 각 메소드는 용도에 맞는 빈을 사용하여 API를 호출합니다.

parknew0 committed:
test(service, controller)

urarik reviewed (approved):
수고 많았습니다~ 전체적으로 깔끔하게 잘 작성해주셨네요! 머지할게요~~
